core/chains/evm/txmgr/orm.go

```diff
@@ -540,3 +575,182 @@ WHERE eth_tx_attempts.state = 'in_progress' AND eth_txes.from_address = $1 AND e
+		err = loadEthTxAttempts(tx, etx)
+		return errors.Wrap(err, "FindEthTxWithNonce failed to load eth_tx_attempts")
+	}, pg.OptReadOnlyTx())
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	}
+	return
+}
+
+func updateEthTxAttemptUnbroadcast(q pg.Queryer, attempt EthTxAttempt) error {
+	if attempt.State != EthTxAttemptBroadcast {
+		return errors.New("expected eth_tx_attempt to be broadcast")
+	}
+	_, err := q.Exec(`UPDATE eth_tx_attempts SET broadcast_before_block_num = NULL, state = 'in_progress' WHERE id = $1`, attempt.ID)
+	return errors.Wrap(err, "updateEthTxAttemptUnbroadcast failed")
+}
+
+func updateEthTxUnconfirm(q pg.Queryer, etx EthTx) error {
+	if etx.State != EthTxConfirmed {
+		return errors.New("expected eth_tx state to be confirmed")
+	}
+	_, err := q.Exec(`UPDATE eth_txes SET state = 'unconfirmed' WHERE id = $1`, etx.ID)
+	return errors.Wrap(err, "updateEthTxUnconfirm failed")
+}
+
+func deleteEthReceipts(q pg.Queryer, etxID int64) (err error) {
+	_, err = q.Exec(`
+DELETE FROM eth_receipts
+USING eth_tx_attempts
+WHERE eth_receipts.tx_hash = eth_tx_attempts.hash
+AND eth_tx_attempts.eth_tx_id = $1
+	`, etxID)
+	return errors.Wrap(err, "deleteEthReceipts failed")
+}
+
+func (o *orm) UpdateEthTxForRebroadcast(etx EthTx, etxAttempt EthTxAttempt) error {
+	return o.q.Transaction(func(tx pg.Queryer) error {
+		if err := deleteEthReceipts(tx, etx.ID); err != nil {
+			return errors.Wrapf(err, "deleteEthReceipts failed for etx %v", etx.ID)
+		}
+		if err := updateEthTxUnconfirm(tx, etx); err != nil {
+			return errors.Wrapf(err, "updateEthTxUnconfirm failed for etx %v", etx.ID)
+		}
+		return updateEthTxAttemptUnbroadcast(tx, etxAttempt)
+	})
+}
+
+func (o *orm) FindTransactionsConfirmedInBlockRange(highBlockNumber, lowBlockNumber int64, chainID big.Int) (etxs []*EthTx, err error) {
+	err = o.q.Transaction(func(tx pg.Queryer) error {
+		err = tx.Select(&etxs, `
+SELECT DISTINCT eth_txes.* FROM eth_txes
+INNER JOIN eth_tx_attempts ON eth_txes.id = eth_tx_attempts.eth_tx_id AND eth_tx_attempts.state = 'broadcast'
+INNER JOIN eth_receipts ON eth_receipts.tx_hash = eth_tx_attempts.hash
+WHERE eth_txes.state IN ('confirmed', 'confirmed_missing_receipt') AND block_number BETWEEN $1 AND $2 AND evm_chain_id = $3
+ORDER BY nonce ASC
+`, lowBlockNumber, highBlockNumber, chainID.String())
+		if err != nil {
+			return errors.Wrap(err, "FindTransactionsConfirmedInBlockRange failed to load eth_txes")
+		}
+		if err = loadEthTxesAttempts(tx, etxs); err != nil {
+			return errors.Wrap(err, "FindTransactionsConfirmedInBlockRange failed to load eth_tx_attempts")
+		}
+		err = loadEthTxesAttemptsReceipts(tx, etxs)
+		return errors.Wrap(err, "FindTransactionsConfirmedInBlockRange failed to load eth_receipts")
+	}, pg.OptReadOnlyTx())
+	return etxs, errors.Wrap(err, "FindTransactionsConfirmedInBlockRange failed")
+}
+
+func (o *orm) saveAttemptWithNewState(q pg.Queryer, timeout time.Duration, attempt EthTxAttempt, broadcastAt time.Time) error {
```

Comment:
this function has remained private

Reply:
I didn't quite follow. Why does it need to stay private?
Can't we generate the Queryer via "q := pg.NewQ(db, lggr, config)" inside the orm?

Reply:
You can't generate the queryer inside because it passes on the queryer to `saveAttemptWithNewState`. in `saveAttemptWithNewState` if it's a `sqlx.Tx` it will add the statements as part of the transaction. So if you always generated the queryer (or just used the orm's existing one) we couldn't compose queries to be part of the same `sqlx.Tx`

Reply:
See https://github.com/smartcontractkit/chainlink/pull/8523/files/8f16fa705146b0faa8331680fd1695f046327850#diff-17acb6da51a93f6407f839210ff4ecd5809bc34eccc3252aa5e613429d4ca386R706 for an example of why it needs to be passed in. 

Short answer is if it's a `sqlx.Tx` we use the outer transaction and if not, then we start a new transaction. If we always create a new transaction we're not able to support to former use case when we want to combine a bunch of statements into one postgres transaction

Reply:
I think it's interesting to explore a future world where all orm functions are standardized to have the ability to optionally be part of same transaction and get passed in a new parent context

Reply:
@augustbleeds we have an existing pattern to do this, it requires every function to take an optional `qopts` at the end. You can pass in contexts and parent txes this way.

Reply:
So the pattern would be in an orm method have optional `qopts` at end

And then in the orm method call?
`qq := o.q.WithOpts(qopts)`

I think that would solve the passing in new contexts in a more elegant way. However, to support transactions (sqlx.Tx) I think some more lift would be required. We'd have to wrap `pg.SqlxTransaction` around every query execution call so statements would be executed in the same postgres transaction. Also, not sure what our approach would be on orm methods with multiple statements not wrapped in one transaction. 

Good to know though for the future if we want to make this a standard.